stko/molecular/molecule_splitter/molecule_splitter.py

```diff
@@ -0,0 +1,157 @@
+        )
+        split_mols = splitter.split(full_mol)
+
+    """
+
+    def __init__(
+        self,
+        breaker_smarts,
+        bond_deleter_ids,
+        replacer_smarts,
+        functional_groups,
+    ):
+        """
+        Initialize a :class:`MoleculeSplitter`.
+
+        Parameters
+        ----------
+        breaker_smarts : :class:`str`
+            SMARTS string to find the substructure to break.
+
+        bond_deleter_ids : :class:`tuple` of :class:`int`
+            Index of atoms in `breaker_smarts` to breka bond between.
+
+        replacer_smarts : :class:`str`
+            SMARTS string of atom to replace dummy atoms with. This
+            must be a single atom.
+
+        functional_groups :
+        :class:`iterable` of :class:`stk.FunctionalGroupFactory`
+            Functional group factories to use to define new building
+            block.
+
+        """
+
+        self._breaker_smarts = breaker_smarts
+        self._bond_deleter_ids = bond_deleter_ids
+        self._replacer_smarts = replacer_smarts
+        self._functional_groups = functional_groups
+
+    def split(self, molecule):
+        """
+        Split a molecule into multiple molecules.
+
+        Parameters
+        ----------
+        molecule : :class:`stk.Molecule`
+            Molecule to split.
+
+        Returns
+        -------
+        :class:`iterable` of :class:`stk.BuildingBlock`
+            The resulting list of building blocks, with new functional
+            groups assigned.
+
+        """
+
+        new_building_blocks = []
+
+        rdkit_mol = molecule.to_rdkit_mol()
+        rdkit.SanitizeMol(rdkit_mol)
+        bond_pair_ids_to_delete = []
+        reactable_atom_ids = []
+        for atom_ids in rdkit_mol.GetSubstructMatches(
+            query=rdkit.MolFromSmarts(self._breaker_smarts)
+        ):
+            translated_bond_atom_ids = (
+                atom_ids[self._bond_deleter_ids[0]],
+                atom_ids[self._bond_deleter_ids[1]],
+            )
+            bond_pair_ids_to_delete.extend(
+                tuple(sorted((i, j)))
+                for i, j in combinations(translated_bond_atom_ids, 2)
+            )
+            reactable_atom_ids.extend(
+                i for i in translated_bond_atom_ids
+            )
+
+        bond_ids_to_delete = []
+        for bond in rdkit_mol.GetBonds():
+            idxs = tuple(sorted((
+                bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+            )))
+            if idxs in bond_pair_ids_to_delete:
+                bond_ids_to_delete.append(bond.GetIdx())
+
+        # Delete bonds and atoms.
+        fragments = rdkit.GetMolFrags(rdkit.FragmentOnBonds(
+            mol=rdkit_mol,
+            bondIndices=tuple(bond_ids_to_delete),
+            addDummies=True,
+        ), asMols=True)
+
+        # Add replacers bonded to * atoms.
+        for i in rdkit.MolFromSmarts(self._replacer_smarts).GetAtoms():
```

Comment:
Why is this in a for loop if it's only being executed once? Maybe call `next` on it instead?

Reply:
Or list